In [1]:
import kagglehub

# Download latest version
path = kagglehub.dataset_download("new-york-city/ny-2015-street-tree-census-tree-data")

print("Path to dataset files:", path)

100%|██████████| 49.8M/49.8M [00:03<00:00, 16.0MB/s]

Extracting files...


Path to dataset files: /root/.cache/kagglehub/datasets/new-york-city/ny-2015-street-tree-census-tree-data/versions/4


In [2]:
import pandas as pd
import numpy as np
pd.set_option('display.max_columns', None)
data = pd.read_csv('/root/.cache/kagglehub/datasets/new-york-city/ny-2015-street-tree-census-tree-data/versions/4/2015-street-tree-census-tree-data.csv')
# data.shape
# data.columns
data.head()

tree_id  block_id               created_at  tree_dbh  stump_diam curb_loc  \
0   180683    348711  2015-08-27T00:00:00.000         3           0   OnCurb   
1   200540    315986  2015-09-03T00:00:00.000        21           0   OnCurb   
2   204026    218365  2015-09-05T00:00:00.000         3           0   OnCurb   
3   204337    217969  2015-09-05T00:00:00.000        10           0   OnCurb   
4   189565    223043  2015-08-30T00:00:00.000        21           0   OnCurb   

  status health                           spc_latin       spc_common steward  \
0  Alive   Fair                         Acer rubrum        red maple     NaN   
1  Alive   Fair                   Quercus palustris          pin oak     NaN   
2  Alive   Good  Gleditsia triacanthos var. inermis      honeylocust    1or2   
3  Alive   Good  Gleditsia triacanthos var. inermis      honeylocust     NaN   
4  Alive   Good                     Tilia americana  American linden     NaN   

  guards  sidewalk         user_type problems root_stone root_grate  \
0    NaN  NoDamage  TreesCount Staff      NaN         No         No   
1    NaN    Damage  TreesCount Staff   Stones        Yes         No   
2    NaN    Damage         Volunteer      NaN         No         No   
3    NaN    Damage         Volunteer   Stones        Yes         No   
4    NaN    Damage         Volunteer   Stones        Yes         No   

  root_other trunk_wire trnk_light trnk_other brch_light brch_shoe brch_other  \
0         No         No         No         No         No        No         No   
1         No         No         No         No         No        No         No   
2         No         No         No         No         No        No         No   
3         No         No         No         No         No        No         No   
4         No         No         No         No         No        No         No   

             address  postcode      zip_city  community board  borocode  \
0  108-005 70 AVENUE     11375  Forest Hills              406         4   
1   147-074 7 AVENUE     11357    Whitestone              407         4   
2  390 MORGAN AVENUE     11211      Brooklyn              301         3   
3  1027 GRAND STREET     11211      Brooklyn              301         3   
4       603 6 STREET     11215      Brooklyn              306         3   

    borough  cncldist  st_assem  st_senate   nta            nta_name  boro_ct  \
0    Queens        29        28         16  QN17        Forest Hills  4073900   
1    Queens        19        27         11  QN49          Whitestone  4097300   
2  Brooklyn        34        50         18  BK90   East Williamsburg  3044900   
3  Brooklyn        34        53         18  BK90   East Williamsburg  3044900   
4  Brooklyn        39        44         21  BK37  Park Slope-Gowanus  3016500   

      state   latitude  longitude         x_sp         y_sp  council district  \
0  New York  40.723092 -73.844215  1027431.148  202756.7687              29.0   
1  New York  40.794111 -73.818679  1034455.701  228644.8374              19.0   
2  New York  40.717581 -73.936608  1001822.831  200716.8913              34.0   
3  New York  40.713537 -73.934456  1002420.358  199244.2531              34.0   
4  New York  40.666778 -73.975979   990913.775  182202.4260              39.0   

   census tract        bin           bbl  
0         739.0  4052307.0  4.022210e+09  
1         973.0  4101931.0  4.044750e+09  
2         449.0  3338310.0  3.028870e+09  
3         449.0  3338342.0  3.029250e+09  
4         165.0  3025654.0  3.010850e+09

In [3]:
data[['tree_dbh', 'stump_diam', 'latitude', 'longitude', 'x_sp', 'y_sp']].describe()

tree_dbh     stump_diam       latitude      longitude  \
count  683788.000000  683788.000000  683788.000000  683788.000000   
mean       11.279787       0.432463      40.701261     -73.924060   
std         8.723042       3.290241       0.090311       0.123583   
min         0.000000       0.000000      40.498466     -74.254965   
25%         4.000000       0.000000      40.631928     -73.980500   
50%         9.000000       0.000000      40.700612     -73.912911   
75%        16.000000       0.000000      40.762228     -73.834910   
max       450.000000     140.000000      40.912918     -73.700488   

               x_sp           y_sp  
count  6.837880e+05  683788.000000  
mean   1.005280e+06  194798.424625  
std    3.428505e+04   32902.061114  
min    9.133493e+05  120973.792200  
25%    9.896578e+05  169515.153700  
50%    1.008386e+06  194560.252500  
75%    1.029991e+06  217019.571950  
max    1.067248e+06  271894.092100

Здесь мы получаем информацию о DataFrame в целом, о каждом столбце и его ненулевых значениях(во втором столбце)

Всего у нас 683788 записей в датафрэйме таким образом мы получаем, что пустые значения, над которыми нужно поработать содержатся в следующих столбцах:

Типа object: health, spc_latin, spc_common, steward, guards, sidewalk, problems

Типа float64: council district, census tract, bin, bbl

In [4]:
data.info(max_cols=44, verbose=True, show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 683788 entries, 0 to 683787
Data columns (total 45 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   tree_id           683788 non-null  int64  
 1   block_id          683788 non-null  int64  
 2   created_at        683788 non-null  object 
 3   tree_dbh          683788 non-null  int64  
 4   stump_diam        683788 non-null  int64  
 5   curb_loc          683788 non-null  object 
 6   status            683788 non-null  object 
 7   health            652172 non-null  object 
 8   spc_latin         652169 non-null  object 
 9   spc_common        652169 non-null  object 
 10  steward           164350 non-null  object 
 11  guards            79866 non-null   object 
 12  sidewalk          652172 non-null  object 
 13  user_type         683788 non-null  object 
 14  problems          225844 non-null  object 
 15  root_stone        683788 non-null  object 
 16  root_grate        68

Подсчитаем все нулевые значения в каждом столбце:

In [5]:
data.isnull().sum()

tree_id                  0
block_id                 0
created_at               0
tree_dbh                 0
stump_diam               0
curb_loc                 0
status                   0
health               31616
spc_latin            31619
spc_common           31619
steward             519438
guards              603922
sidewalk             31616
user_type                0
problems            457944
root_stone               0
root_grate               0
root_other               0
trunk_wire               0
trnk_light               0
trnk_other               0
brch_light               0
brch_shoe                0
brch_other               0
address                  0
postcode                 0
zip_city                 0
community board          0
borocode                 0
borough                  0
cncldist                 0
st_assem                 0
st_senate                0
nta                      0
nta_name                 0
boro_ct                  0
state                    0
latitude                 0
longitude                0
x_sp                     0
y_sp                     0
council district      6519
census tract          6519
bin                   9559
bbl                   9559
dtype: int64

Посчитаем количество живых деревьев, пней и мертвых деревьев:

In [6]:
data['status'].value_counts()

status
Alive    652173
Stump     17654
Dead      13961
Name: count, dtype: int64

Поскольку наши наблюдения о состоянии дерева сущетвуют только для живых деревьев status==Alive (те которые не являются пнями или мертвыми), то оставим в датафрэйме только живые деревья.

В сумме неживых деревьев у нас 31615

In [7]:

df_alive=data[data["status"]=="Alive"]
df_alive['status'].value_counts()

status
Alive    652173
Name: count, dtype: int64

Снова подсчитаем количество пустых значений в каждом из столбцов:

In [8]:
df_alive.isnull().sum()

tree_id                  0
block_id                 0
created_at               0
tree_dbh                 0
stump_diam               0
curb_loc                 0
status                   0
health                   1
spc_latin                5
spc_common               5
steward             487823
guards              572307
sidewalk                 1
user_type                0
problems            426329
root_stone               0
root_grate               0
root_other               0
trunk_wire               0
trnk_light               0
trnk_other               0
brch_light               0
brch_shoe                0
brch_other               0
address                  0
postcode                 0
zip_city                 0
community board          0
borocode                 0
borough                  0
cncldist                 0
st_assem                 0
st_senate                0
nta                      0
nta_name                 0
boro_ct                  0
state                    0
latitude                 0
longitude                0
x_sp                     0
y_sp                     0
council district      6206
census tract          6206
bin                   9103
bbl                   9103
dtype: int64

Поскольку нам необходимо, чтобы все labels были непустыми, то удалим 1 пустое значение из health.

А также удалим пустые значения из spc_latin, sidewalk

In [9]:
df_alive=df_alive[df_alive['health'].notna()]
df_alive=df_alive[df_alive['spc_latin'].notna()]
df_alive=df_alive[df_alive['sidewalk'].notna()]
df_alive['health'] = df_alive['health'].replace({'Good': 1, 'Fair': 2, 'Poor': 3})
df_alive.isnull().sum()

<ipython-input-9-afe532b91d53>:4: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df_alive['health'] = df_alive['health'].replace({'Good': 1, 'Fair': 2, 'Poor': 3})


tree_id                  0
block_id                 0
created_at               0
tree_dbh                 0
stump_diam               0
curb_loc                 0
status                   0
health                   0
spc_latin                0
spc_common               0
steward             487819
guards              572300
sidewalk                 0
user_type                0
problems            426324
root_stone               0
root_grate               0
root_other               0
trunk_wire               0
trnk_light               0
trnk_other               0
brch_light               0
brch_shoe                0
brch_other               0
address                  0
postcode                 0
zip_city                 0
community board          0
borocode                 0
borough                  0
cncldist                 0
st_assem                 0
st_senate                0
nta                      0
nta_name                 0
boro_ct                  0
state                    0
latitude                 0
longitude                0
x_sp                     0
y_sp                     0
council district      6206
census tract          6206
bin                   9103
bbl                   9103
dtype: int64

Далее удалим столбцы "council district", "census tract" из датафрэйма, поскольку "council district" полностью повторяет "cncldist", а "census tract" полностью содержится в "boro_ct"

In [10]:
df_alive=df_alive.drop(columns = ['council district', 'census tract'], axis = 1)

Заменим пустые значения на None, как замечено в таблице, для них нет охраны, поэтому в дальнейшем - это будет отдельное значение - 0.

Аналогично сделаем для problems и steward

In [11]:
df_alive['guards'] = df_alive['guards'].fillna("None")
df_alive['guards'].value_counts()
df_alive['steward'] = df_alive['steward'].fillna("None")
df_alive['steward'].value_counts()
df_alive['problems'] = df_alive['problems'].fillna("None")
df_alive['problems'].value_counts()
df_alive['steward'] = df_alive['steward'].replace({'1or2': 1, '3or4': 2, '4orMore': 3, 'None': 0})
df_alive['guards'] = df_alive['guards'].replace({'Harmful': 1, 'Helpful': 2, 'Unsure': 3, 'None': 0})

df_alive.isnull().sum()

<ipython-input-11-1a4bd5a72c46>:7: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df_alive['steward'] = df_alive['steward'].replace({'1or2': 1, '3or4': 2, '4orMore': 3, 'None': 0})
<ipython-input-11-1a4bd5a72c46>:8: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df_alive['guards'] = df_alive['guards'].replace({'Harmful': 1, 'Helpful': 2, 'Unsure': 3, 'None': 0})


tree_id               0
block_id              0
created_at            0
tree_dbh              0
stump_diam            0
curb_loc              0
status                0
health                0
spc_latin             0
spc_common            0
steward               0
guards                0
sidewalk              0
user_type             0
problems              0
root_stone            0
root_grate            0
root_other            0
trunk_wire            0
trnk_light            0
trnk_other            0
brch_light            0
brch_shoe             0
brch_other            0
address               0
postcode              0
zip_city              0
community board       0
borocode              0
borough               0
cncldist              0
st_assem              0
st_senate             0
nta                   0
nta_name              0
boro_ct               0
state                 0
latitude              0
longitude             0
x_sp                  0
y_sp                  0
bin                9103
bbl                9103
dtype: int64

Теперь заменим в столбцах Yes на 1 No на 0, для столбцов root_stone, root_grate, root_other, trunk_wire, trnk_light, trnk_other, brch_light, brch_shoe, brch_other, поскольку для модели важно получить численные значения на вход.

И также разобьем по категориям user_type, sidewalk, curb_loc

In [12]:
df_alive['root_stone'] = df_alive['root_stone'].replace({'Yes': 1, 'No': 0})
df_alive['root_grate'] = df_alive['root_grate'].replace({'Yes': 1, 'No': 0})
df_alive['root_other'] = df_alive['root_other'].replace({'Yes': 1, 'No': 0})
df_alive['trunk_wire'] = df_alive['trunk_wire'].replace({'Yes': 1, 'No': 0})
df_alive['trnk_light'] = df_alive['trnk_light'].replace({'Yes': 1, 'No': 0})
df_alive['trnk_other'] = df_alive['trnk_other'].replace({'Yes': 1, 'No': 0})
df_alive['brch_light'] = df_alive['brch_light'].replace({'Yes': 1, 'No': 0})
df_alive['brch_shoe'] = df_alive['brch_shoe'].replace({'Yes': 1, 'No': 0})
df_alive['brch_other'] = df_alive['brch_other'].replace({'Yes': 1, 'No': 0})
df_alive['curb_loc'] = df_alive['curb_loc'].replace({'OnCurb': 1, 'OffsetFromCurb': 0})
df_alive['sidewalk'] = df_alive['sidewalk'].replace({'Damage': 1, 'NoDamage': 0})
df_alive['user_type'] = df_alive['user_type'].replace({'Volunteer': 1, 'TreesCount Staff': 2, 'NYC Parks Staff': 3})
df_alive['user_type'].value_counts()

<ipython-input-12-c59d4eb9d416>:1: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df_alive['root_stone'] = df_alive['root_stone'].replace({'Yes': 1, 'No': 0})
<ipython-input-12-c59d4eb9d416>:2: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df_alive['root_grate'] = df_alive['root_grate'].replace({'Yes': 1, 'No': 0})
<ipython-input-12-c59d4eb9d416>:3: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. T

user_type
2    282254
1    207776
3    162136
Name: count, dtype: int64

Также по категориям разобьем деревья, относящиеся к разным видам, виды будем определять по их латинским названиям spc_latin.

In [13]:
unique_values = df_alive['spc_latin'].unique()

print(unique_values.shape)
df_alive['spc_latin_values'], unique_values = pd.factorize(df_alive['spc_latin'])
df_alive[['spc_latin_values', 'spc_latin']]

(132,)


spc_latin_values                           spc_latin
0                      0                         Acer rubrum
1                      1                   Quercus palustris
2                      2  Gleditsia triacanthos var. inermis
3                      2  Gleditsia triacanthos var. inermis
4                      3                     Tilia americana
...                  ...                                 ...
683783                 1                   Quercus palustris
683784                35                 Cladrastis kentukea
683785                 0                         Acer rubrum
683786                 0                         Acer rubrum
683787                 0                         Acer rubrum

[652166 rows x 2 columns]

Удалим некоторое число записей о живых деревьях с состоянием Good, где нет видимых проблем и они являются представителями самых распространенных видов.

В дальнейшем опишем почему был сделан этот шаг: из-за значительного перекоса в данных в сторону значений Good

In [14]:
filtered_df = df_alive[(df_alive['problems'] == 'None') & (df_alive['health'] == 1)]

# Подсчитываем уникальные значения в spc_latin_values для отфильтрованных данных
filtered_df['spc_latin_values'].value_counts()

spc_latin_values
4      37948
25     34971
2      32312
1      27120
40     20623
       ...  
130       23
107       22
110       17
116       14
124        7
Name: count, Length: 132, dtype: int64

In [15]:
df_alive['spc_latin_values'].value_counts()

spc_latin_values
4      87014
2      64262
25     58931
1      53185
17     34189
       ...  
116       37
130       33
107       29
110       25
124       10
Name: count, Length: 132, dtype: int64

In [16]:
filtered_df_1 = df_alive[(df_alive['problems'] == 'None') & (df_alive['spc_latin_values'] == 4) & (df_alive['health'] == 1)]
filtered_df_2 = df_alive[(df_alive['problems'] == 'None') & (df_alive['spc_latin_values'] == 2) & (df_alive['health'] == 1)]
filtered_df_3 = df_alive[(df_alive['problems'] == 'None') & (df_alive['spc_latin_values'] == 25) & (df_alive['health'] == 1)]
filtered_df_4 = df_alive[(df_alive['problems'] == 'None') & (df_alive['spc_latin_values'] == 1) & (df_alive['health'] == 1)]

num_to_remove_max = 35000
num_to_remove_min = 20000

# Получаем индексы строк для удаления
indices_to_remove_1 = filtered_df_1.head(num_to_remove_max).index
indices_to_remove_2 = filtered_df_2.head(num_to_remove_max).index
indices_to_remove_3 = filtered_df_3.head(num_to_remove_min).index
indices_to_remove_4 = filtered_df_4.head(num_to_remove_min).index
df_alive = df_alive.drop(indices_to_remove_1)
df_alive = df_alive.drop(indices_to_remove_2)
df_alive = df_alive.drop(indices_to_remove_3)
df_alive = df_alive.drop(indices_to_remove_4)
df_alive['spc_latin_values'].value_counts()

spc_latin_values
4      52014
25     38931
17     34189
1      33185
2      31950
       ...  
116       37
130       33
107       29
110       25
124       10
Name: count, Length: 132, dtype: int64

Мы не берем 'problems', поскольку значения, содержащиеся в ячейке полностью описываются в столбцах 'root_stone', 'root_grate', 'root_other', 'trunk_wire', 'trnk_light', 'trnk_other', 'brch_light', 'brch_shoe', 'brch_other'.

stump_diam не берем поскольку это значение не относится к живым деревьям, только к пням.

А boroname не берем поскольку он полностью совпадает с borocode.

In [17]:
columns_to_keep = ['tree_dbh',
                   'curb_loc',
                   'spc_latin_values',
                   'steward',
                   'guards',
                   'sidewalk',
                   'user_type',
                   'root_stone',
                   'root_grate',
                   'root_other',
                   'trunk_wire',
                   'trnk_light',
                   'trnk_other',
                   'brch_light',
                   'brch_shoe',
                   'brch_other',
                   'postcode',
                   'community board',
                   'borocode',
                   'cncldist',
                   'st_assem',
                   'st_senate',
                   'boro_ct',
                   'latitude',
                   'longitude',
                   'x_sp',
                   'y_sp']

Здесь описан перекос в данных, более 70% данных у нас относятся к классу Good, и меньше 5% - к классу Poor.

In [18]:
X = df_alive[columns_to_keep]
y = df_alive['health']
y
df_alive['health'].value_counts()

health
1    421533
2     96504
3     26817
Name: count, dtype: int64

Далее делаем train и test выборки в соотношении 80%/20%

In [19]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report
from sklearn.metrics import f1_score

X_train, X_test, y_train, y_test = train_test_split(X, y, stratify = y, random_state = 42)
y_train.value_counts()

health
1    316149
2     72378
3     20113
Name: count, dtype: int64

In [20]:
y_train.value_counts(normalize=True)

health
1    0.773661
2    0.177119
3    0.049219
Name: proportion, dtype: float64

Далее делаем under-сэмплинг для train датасета, уменьшаем количество примеров с из класса Good, для того, чтобы немного сбалансировать данные

In [21]:
from imblearn.over_sampling import SMOTE
from imblearn.under_sampling import RandomUnderSampler
# Применяем SMOTE только к обучающим данным (X_train, y_train)
under_sampler = RandomUnderSampler(sampling_strategy={1: 200000, 2: 63040, 3: 18111}, random_state=42)
X_resampled, y_resampled = under_sampler.fit_resample(X_train, y_train)
y_resampled.value_counts()




health
1    200000
2     63040
3     18111
Name: count, dtype: int64

Далее применяем метод SMOTE для синтетического увеличения записей Over-sampling, чтобы сбалансировать датасет, то есть увеличим количество записей для классов Fair и Poor

In [22]:
smote = SMOTE(sampling_strategy={1: 200000, 2: 100000, 3: 50000}, random_state=42)
X_train_resampled, y_train_resampled = smote.fit_resample(X_resampled, y_resampled)

print("Resampled class distribution in train set:", dict(zip(*np.unique(y_train_resampled, return_counts=True))))
y_train_resampled.value_counts()

Resampled class distribution in train set: {1: 200000, 2: 100000, 3: 50000}


health
1    200000
2    100000
3     50000
Name: count, dtype: int64

Далее рассмотрим классификацию с помощью RandomForest классификатора, для того, чтобы в дальнейшем узнать, каких результатов в сравнении мы сможем добиться, используя нейронную сеть для классификации.

Для оценки будем использовать не accuracy, а f1-score, поскольку мы имеем значительный перекос в данных и метрика accuracy не будет достаточно информативной для нас.

In [23]:
rfc_2_smote = RandomForestClassifier(n_estimators = 100, class_weight='balanced', max_depth = 20, random_state = 42)
rfc_2_smote.fit(X_train_resampled, y_train_resampled)
y_pred_2_smote = rfc_2_smote.predict(X_test)

print(classification_report(y_test, y_pred_2_smote))
print("F1-Score:", f1_score(y_test, y_pred_2_smote, average='weighted'))

              precision    recall  f1-score   support

           1       0.86      0.80      0.83    105384
           2       0.41      0.40      0.41     24126
           3       0.15      0.34      0.21      6704

    accuracy                           0.71    136214
   macro avg       0.47      0.51      0.48    136214
weighted avg       0.75      0.71      0.72    136214

F1-Score: 0.7237462342114466


В сравнении посмотрим, насколько наш классификатор с использованием under-сэмплинга и over-сэмплинга лучше базового классификатора

In [ ]:
rfc_2 = RandomForestClassifier(n_estimators = 100, class_weight='balanced', max_depth = 20, random_state = 42)
rfc_2.fit(X_train, y_train)
rfc_2.score(X_train, y_train), rfc_2.score(X_test, y_test)

In [25]:
from sklearn.metrics import classification_report
y_pred_2 = rfc_2.predict(X_test)
print(classification_report(y_test, y_pred_2))

              precision    recall  f1-score   support

           1       0.86      0.90      0.88    105384
           2       0.48      0.39      0.43     24126
           3       0.29      0.28      0.29      6704

    accuracy                           0.78    136214
   macro avg       0.54      0.52      0.53    136214
weighted avg       0.76      0.78      0.77    136214



Далее для нашей задачи была выбрана нейронная сеть с архитектурой Multilayer Perceptron, поскольку она довольно проста и хорошо подходит для решения задач классификации, в особенности многоклассовой, как в нашем случае. Также она хорошо себя показывает при работе с табличными данными.

Было выбрано 3 слоя, поскольку задача несложная и большее количество слоев могло бы привести к переобучению. Первый слой состоит из 256 нейронов и учитывает базовые признаки в данных. Второй слой содержит 128 нейронов и позволяет учитывать более сложные комбинации признаков. Последний - 3 слой на выходе дает 3 класса, поскольку у нас 3 класса.

Для данной задачи были выбраны learning rate = 0.0001 и также был использован lr_scheduler, такой начальный lr был выбран путем экспериментов и показал наиболее плавный процесс обучения.

Weight decay был выбран равным 0.00001, для борьбы с переобучением.
Также был добавлен dropout = 0.3 для первого слоя и 0.2 для второго.

Количество шагов тренировки было выбрано равным 100, поскольку у нас довольно маленький learning rate это позволяет модели не сильно переобучиться, но при этом выявить более точные зависимости при дисбалансе классов.

Также для решения проблемы с дисбалансом классов была выбрана вместо классической функции потерь - взвешенная функция потерь(кросс-энтропия).

In [26]:
import torch
import torch.nn as nn
import torch.optim as optim
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.utils import class_weight
from sklearn.metrics import classification_report
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import f1_score
from sklearn.utils.class_weight import compute_class_weight
from imblearn.under_sampling import RandomUnderSampler
from imblearn.over_sampling import SMOTE

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print(f"Using device: {device}")

Using device: cuda


In [30]:
class MLP(nn.Module):
    def __init__(self, input_dim, num_classes):
        super(MLP, self).__init__()
        self.fc1 = nn.Linear(input_dim, 256)
        self.bn1 = nn.BatchNorm1d(256)
        self.dropout1 = nn.Dropout(0.3)

        self.fc2 = nn.Linear(256, 128)
        self.bn2 = nn.BatchNorm1d(128)
        self.dropout2 = nn.Dropout(0.2)

        self.fc3 = nn.Linear(128, num_classes)

    def forward(self, x):
        x = torch.relu(self.bn1(self.fc1(x)))
        x = self.dropout1(x)

        x = torch.relu(self.bn2(self.fc2(x)))
        x = self.dropout2(x)

        x = self.fc3(x)
        return x


X = df_alive[columns_to_keep].values
y = df_alive['health'].values
y = y - 1 # приводим наши метки к диапазону от 0 до 2, вместо 1 до 3
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

under_sampler = RandomUnderSampler(sampling_strategy={0: 200000, 1: 63040, 2: 18111}, random_state=42)
X_resampled, y_resampled = under_sampler.fit_resample(X_train, y_train)
print(pd.Series(y_resampled).value_counts())
smote = SMOTE(sampling_strategy={0: 200000, 1: 150000, 2: 100000}, random_state=42)
X_train_resampled, y_train_resampled = smote.fit_resample(X_resampled, y_resampled)
print(pd.Series(y_train_resampled).value_counts())

device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
X_train_tensor = torch.tensor(X_train_resampled, dtype=torch.float32).to(device)
X_test_tensor = torch.tensor(X_test, dtype=torch.float32).to(device)
y_train_tensor = torch.tensor(y_train_resampled, dtype=torch.long).to(device)
y_test_tensor = torch.tensor(y_test, dtype=torch.long).to(device)

# Поскольку у нас присутствует дисбаланс классов будем использовать функцию потерь(кросс-энтропия) с весами классов
class_weights = compute_class_weight('balanced', classes=np.unique(y_train_resampled), y=y_train_resampled)
class_weights = torch.tensor(class_weights, dtype=torch.float32).to(device)

model = MLP(input_dim=X_train.shape[1], num_classes=3).to(device)

criterion = nn.CrossEntropyLoss(weight=class_weights)  # Взвешенная кросс-энтропия
optimizer = optim.Adam(model.parameters(), lr=0.0001, weight_decay=0.00001)
scheduler = optim.lr_scheduler.ReduceLROnPlateau(optimizer, 'min', patience=5, factor=0.5)

epochs = 100
batch_size = 32

for epoch in range(epochs):
  # Тренировка модели
  model.train()
  permutation = torch.randperm(X_train_tensor.size()[0])

  running_loss = 0.0
  for i in range(0, X_train_tensor.size()[0], batch_size):
    indices = permutation[i:i + batch_size]
    batch_x, batch_y = X_train_tensor[indices], y_train_tensor[indices]

    optimizer.zero_grad()

    outputs = model(batch_x)
    loss = criterion(outputs, batch_y)

    # Обратный проход
    loss.backward()
    optimizer.step()

    running_loss += loss.item()

  # Оценка модели
  model.eval()
  with torch.no_grad():
    # Прогноз на обучающей выборке
    y_train_pred_prob = model(X_train_tensor)
    _, y_train_pred = torch.max(y_train_pred_prob, 1)

    # Прогноз на тестовой выборке
    y_test_pred_prob = model(X_test_tensor)
    _, y_test_pred = torch.max(y_test_pred_prob, 1)

    # Метрики на обучающей выборке
    train_f1 = f1_score(y_train_tensor.cpu().numpy(), y_train_pred.cpu().numpy(), average='weighted')
    train_report = classification_report(y_train_tensor.cpu().numpy(), y_train_pred.cpu().numpy(), output_dict=True, zero_division=0)

    # Метрики на тестовой выборке
    test_f1 = f1_score(y_test_tensor.cpu().numpy(), y_test_pred.cpu().numpy(), average='weighted')
    test_report = classification_report(y_test_tensor.cpu().numpy(), y_test_pred.cpu().numpy(), output_dict=True, zero_division=0)
  print(f'Epoch [{epoch + 1}/{epochs}], Loss: {running_loss / len(permutation):.10f}')
  print(f'Train F1-Score: {train_f1:.6f}')
  print(f'Test F1-Score: {test_f1:.6f}')
  if (epoch+1) % 5 == 0:

    print(f'--- Train Classification Report ---')
    print(train_report)
    print(f'--- Test Classification Report ---')
    print(test_report)
  print('-----------------------------------')

torch.save({
    'epoch': epoch,
    'model_state_dict': model.state_dict(),
    'optimizer_state_dict': optimizer.state_dict(),
    'loss': running_loss,
}, 'MLP_256_128_lr0.0001_lr_scdlr_big.pth')

0    200000
1     63040
2     18111
Name: count, dtype: int64
0    200000
1    150000
2    100000
Name: count, dtype: int64
Epoch [1/100], Loss: 0.0344272255
Train F1-Score: 0.368362
Test F1-Score: 0.578909
-----------------------------------
Epoch [2/100], Loss: 0.0342177233
Train F1-Score: 0.376513
Test F1-Score: 0.608995
-----------------------------------
Epoch [3/100], Loss: 0.0341740813
Train F1-Score: 0.373201
Test F1-Score: 0.609637
-----------------------------------
Epoch [4/100], Loss: 0.0341572327
Train F1-Score: 0.366236
Test F1-Score: 0.615857
-----------------------------------
Epoch [5/100], Loss: 0.0341459899
Train F1-Score: 0.382872
Test F1-Score: 0.583423
--- Train Classification Report ---
{'0': {'precision': 0.48961686043181946, 'recall': 0.64119, 'f1-score': 0.5552450126105453, 'support': 200000.0}, '1': {'precision': 0.3629697995180229, 'recall': 0.14158, 'f1-score': 0.20370343726170093, 'support': 150000.0}, '2': {'precision': 0.2718558992406001, 'recall': 0.352